webhook.go

```diff
@@ -0,0 +1,178 @@
+package webhook
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+
+	kwhhttp "github.com/slok/kubewebhook/pkg/http"
+	kwhlog "github.com/slok/kubewebhook/pkg/log"
+	kwhmutating "github.com/slok/kubewebhook/pkg/webhook/mutating"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// binVolumeName is the name of the volume where the SecretHub CLI binary is stored.
+	binVolumeName = "secrethub-bin"
+
+	// binVolumeMountPath is the mount path where the SecretHub CLI binary can be found.
+	binVolumeMountPath = "/secrethub/bin/"
+)
+
+// binVolume is the shared, in-memory volume where the SecretHub CLI binary lives.
+var binVolume = corev1.Volume{
+	Name: binVolumeName,
+	VolumeSource: corev1.VolumeSource{
+		EmptyDir: &corev1.EmptyDirVolumeSource{
+			Medium: corev1.StorageMediumMemory,
+		},
+	},
+}
+
+// binVolumeMount is the shared volume mount where the SecretHub CLI binary lives.
+var binVolumeMount = corev1.VolumeMount{
+	Name:      binVolumeName,
+	MountPath: binVolumeMountPath,
+	ReadOnly:  true,
+}
+
+// SecretHubMutator is a mutator.
+type SecretHubMutator struct {
+	logger kwhlog.Logger
+}
+
+// Mutate implements MutateFunc and provides the top-level entrypoint for object
+// mutation.
+func (m *SecretHubMutator) Mutate(ctx context.Context, obj metav1.Object) (bool, error) {
+	m.logger.Infof("calling mutate")
+
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return false, nil
+	}
+
+	version, enabled := pod.Annotations["secrethub"]
```

Comment:
Is this annotation specific/clear enough? I feel like we're leaving very little wiggle-room for making changes in the future.

For example, what I could imagine is something like:
```
secrethub/inject: true
secrethub/version: v0.1.2.3
```
Or
```
secrethub/inject/enabled: true
secrethub/inject/version: v0.1.2.3
```

This also allows leaving out the version, in which case the latest version is pulled.

Reply:
Or:
```
secrethub/inject/myapp: true
secrethub/inject/nginx: false
```

That would solve this https://github.com/secrethub/secrethub-kubernetes-mutating-webhook/pull/1/files#r435796044

Because we can then leave out this check:

https://github.com/secrethub/secrethub-kubernetes-mutating-webhook/blob/a89f26ee2f62e17be2017bbae85911580be4570f/webhook.go#L118

Reply:
Yes, allowing for more wiggle-room would make it easier/possible to keep backwards-compatibility in the future. :+1: I'll think about a way to achieve this.

> This also allows leaving out the version, in which case the latest version is pulled.

Is this a use case we want to support or encourage?

I think I'd prefer the idempotency over automatically updating to the latest version. Especially since automatically updating would mean we can never do a major release of the CLI that includes breaking changes, without breaking deployments. Even though we've kept the CLI backwards compatible for a long time now, I'd prefer to keep the option open to ever do a major release. Eventually, there might come a time when we'll need this.
Having the annotation with the version allows users to time the updates of the CLI.

Reply:
Given the scenario of using external factors, such as a `secrethub.env` file you've described below, we may want to enable the mutation on a per-container basis instead of per-pod. What do you think?

Reply:
I like your proposal @florisvdg :+1:

Only thing I'd change is `inject`, as this may be confused with the behavior of the `secrehtub inject` command. How about `secrethub/run/my-app: true`?

Reply:
```
secrethub/inject/myapp: true
secrethub/inject/nginx: false
```
I don't know if varying annotation names are idiomatic for k8s. Do you have an example of this happening somewhere else?

---

We could also do something like:
```
secrethub/inject/include: myapp
```
And
```
secrethub/inject/exclude: nginx
```

Where the default is "run on all containers", unless one of these two is specified, then it's :"all but listed" or "only listed".

Reply:
> I don't know if varying annotation names are idiomatic for k8s

Good point.

> Only thing I'd change is inject, as this may be confused with the behavior of the `secrethub inject` command. How about `secrethub/run/my-app: true`?

True, alternative: 
```
secrethub/mutate/include: myapp1,myapp2
```